flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraSinkBaseTest.java

```diff
@@ -0,0 +1,469 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import static org.mockito.Mockito.mock;
+import static org.powermock.api.mockito.PowerMockito.when;
+
+/**
+ * Tests for the {@link CassandraSinkBase}.
+ */
+public class CassandraSinkBaseTest {
+
+	private static final Logger LOG = LoggerFactory.getLogger(CassandraSinkBaseTest.class);
+
+	private static final String DUMMY_QUERY_STMT = "CQL_Dummy_Stmt";
+
+	private static final String DUMMY_MESSAGE = "Dummy_msg";
+
+	private static final int MAX_THREAD_NUM = 3;
+
+	private static final int MAX_THREAD_POOL_SIZE = 2;
+
+	@BeforeClass
+	public static void doSetUp() throws Exception {
+
+	}
+
+	@BeforeClass
+	public static void doTearDown() throws Exception {
+
+	}
+
+	/**
+	 * Test ensures a NoHostAvailableException would be thrown if a contact point added does not exist.
+	 */
+	@Test(expected = NoHostAvailableException.class)
+	public void testCasHostNotFoundErrorHandling() throws Exception {
+		CassandraSinkBase base = new DummyCassandraSinkBase<>(new ClusterBuilder() {
+			@Override
+			protected Cluster buildCluster(Cluster.Builder builder) {
+				return builder
+					.addContactPoint("127.0.0.1")
+					.withoutJMXReporting()
+					.withoutMetrics().build();
+			}
+		});
+
+		base.open(new Configuration());
+		base.close();
+	}
+
+
+	///////////////////////
+	// Single Thread Test
+	///////////////////////
+
+	/**
+	 * Test ensures the message could be delivered successfully to sink.
+	 */
+	@Test
+	public void testSimpleSuccessfulPath() throws Exception {
+		ClusterBuilder builder = mock(ClusterBuilder.class);
+		MockCassandraSinkBase casSinkFunc = new MockCassandraSinkBase<>(builder);
+		casSinkFunc.setFlushOnCheckpoint(true);
+		casSinkFunc.open(new Configuration());
+
+		casSinkFunc.invokeWithImmediateResult(DUMMY_MESSAGE, PredeterminedResult.IMMEDIATE_SUCCESS);
+
+		casSinkFunc.close();
+
+		//Final pending updates should be zero
+		Assert.assertEquals(0, casSinkFunc.getNumOfPendingRecords());
+	}
+
+	/**
+	 * Test ensures that an asyncError would be thrown on close() if previously message delivery failed.
+	 */
+	@Test
+	public void testAsyncErrorThrownOnClose() throws Exception {
+		ClusterBuilder builder = mock(ClusterBuilder.class);
+		MockCassandraSinkBase casSinkFunc = new MockCassandraSinkBase<>(builder);
+		casSinkFunc.setFlushOnCheckpoint(true);
+
+		casSinkFunc.open(new Configuration());
+
+		casSinkFunc.invokeWithImmediateResult(DUMMY_MESSAGE, PredeterminedResult.IMMEDIATE_FAILURE);
+		try {
+
+			casSinkFunc.close();
+		} catch (IOException e) {
+			//expected async error from close()
+
+			Assert.assertTrue(e.getMessage().contains("Error while sending value"));
+
+			//Final pending updates should be zero
+			Assert.assertEquals(0, casSinkFunc.getNumOfPendingRecords());
+
+			//done
+			return;
+		}
+
+		Assert.fail();
```

Comment:
we usually put the fail within the ```try``` block  (also applies to other tests)

Reply:
That is right. Thanks.